### DISCUSSION TITLE
WaveShare xnucleo_f411re: Error: ** Unable to reset target **

### DISCUSSION THREAD

**Discussion by s163555** [zephyrproject-rtos/zephyr#53460] [Q&A]
**Describe the bug**
Similar to #50590  I am experiencing an issue with resetting the STM32F411RE with west and openocd. I am using the board XNUCLEO-F411RE (an improved NUCLEO-F411RE board from WaveShare). The bug is easily reproduced by flashing zephyr unto the board once, and then after this, the board is unable to reset. 
If the on-board reset button is physically pressed and released with the right timing, the flashing will succeed. If not, it will give the error message:
Error: timed out while waiting for target halted
embedded:startup.tcl:1070: Error: ** Unable to reset target **

Please also mention any information which could help others to understand
the problem you're facing:
 - nucleo_f411re (xnucleo_f411re)
 - I have tried two programmers STLINK V2mini and STLINK V3set, both work can reset with STM32CubeProgrammer 
 - Is this a regression? To my knowledge this has been an issue since I started trying to flash to the board

**To Reproduce**
1. mkdir build; cd build
2. cmake -DBOARD=board\_xyz
3. make

**Impact**
An annoyance, but a significant annoyance since every flashing requires multiple times if I cannot time the button release at the right time.

**Logs and console output**
zephyr SDK 15.2
```
C:\Users\Jeppe\zephyrproject\zephyr>west flash
-- west flash: rebuilding
ninja: no work to do.
-- west flash: using runner openocd
-- runners.openocd: Flashing file: C:/Users/Jeppe/zephyrproject/zephyr/build/zephyr/zephyr.hex
Open On-Chip Debugger 0.12.0-rc3 (2022-12-20-12:54)
Licensed under GNU GPL v2
For bug reports, read
        http://openocd.org/doc/doxygen/bugs.html
Info : The selected transport took over low-level target control. The results might differ compared to plain JTAG/SWD
Info : clock speed 2000 kHz
Info : STLINK V2J40S7 (API v2) VID:PID 0483:3748
Info : Target voltage: 3.289186
Info : [stm32f4x.cpu] Cortex-M4 r0p1 processor detected
Info : [stm32f4x.cpu] target has 6 breakpoints, 4 watchpoints
Info : starting gdb server for stm32f4x.cpu on 3333
Info : Listening on port 3333 for gdb connections
    TargetName         Type       Endian TapName            State
--  ------------------ ---------- ------ ------------------ ------------
 0* stm32f4x.cpu       hla_target little stm32f4x.cpu       running

Info : Unable to match requested speed 2000 kHz, using 1800 kHz
Info : Unable to match requested speed 2000 kHz, using 1800 kHz
Error: timed out while waiting for target halted
TARGET: stm32f4x.cpu - Not halted
FATAL ERROR: command exited with status 1: 'C:/Program Files/OpenOCD/bin/openocd.exe' -s 'C:/Users/Jeppe/zephyrproject/zephyr/boards/arm/nucleo_f411re\support' -s C:/Users/Jeppe/zephyr-sdk-0.15.2/sysroots/AMD64-pokysdk-linux/usr/share/openocd/scripts -f 'C:/Users/Jeppe/zephyrproject/zephyr/boards/arm/nucleo_f411re\support\openocd.cfg' '-c init' '-c targets' -c 'reset init' -c 'flash write_image erase C:/Users/Jeppe/zephyrproject/zephyr/build/zephyr/zephyr.hex' -c 'reset init' -c 'verify_image C:/Users/Jeppe/zephyrproject/zephyr/build/zephyr/zephyr.hex' -c 'reset run' -c shutdown
```

platformIO (zephyr v. 3.2.0)
```
Processing nucleo_f411re (platform: ststm32; framework: zephyr; board: nucleo_f411re)
---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
Verbose mode can be enabled via `-v, --verbose` option
CONFIGURATION: https://docs.platformio.org/page/boards/ststm32/nucleo_f411re.html
PLATFORM: ST STM32 (15.4.1) > ST Nucleo F411RE
HARDWARE: STM32F411RET6 100MHz, 128KB RAM, 512KB Flash
DEBUG: Current (stlink) On-board (stlink) External (blackmagic, cmsis-dap, jlink)
PACKAGES:
 - framework-zephyr @ 2.20500.210226+sha.9f9f3e2f
 - tool-cmake @ 3.21.3
 - tool-dfuutil @ 1.11.0
 - tool-dtc @ 1.4.7
 - tool-ninja @ 1.9.0
 - tool-openocd @ 2.1100.211028 (11.0)
 - tool-stm32duino @ 1.0.2
 - toolchain-gccarmnoneeabi @ 1.80201.181220 (8.2.1)
LDF: Library Dependency Finder -> https://bit.ly/configure-pio-ldf
LDF Modes: Finder ~ chain, Compatibility ~ soft
Found 0 compatible libraries
Scanning dependencies...
No dependencies
Building in release mode
Checking size .pio\build\nucleo_f411re\firmware.elf
Advanced Memory Usage is available via "PlatformIO Home > Project Inspect"
RAM:   [          ]   3.4% (used 4437 bytes from 131072 bytes)
Flash: [          ]   2.6% (used 13636 bytes from 524288 bytes)
Configuring upload protocol...
AVAILABLE: blackmagic, cmsis-dap, jlink, mbed, stlink
CURRENT: upload_protocol = stlink
Uploading .pio\build\nucleo_f411re\firmware.elf
xPack OpenOCD x86_64 Open On-Chip Debugger 0.11.0+dev (2021-10-16-21:19)
Licensed under GNU GPL v2
For bug reports, read
        http://openocd.org/doc/doxygen/bugs.html
debug_level: 1

srst_only separate srst_nogate srst_open_drain connect_deassert_srst

Error: timed out while waiting for target halted
embedded:startup.tcl:1070: Error: ** Unable to reset target **
in procedure 'program'
in procedure 'program_error' called at file "embedded:startup.tcl", line 1107
at file "embedded:startup.tcl", line 1070
*** [upload] Error 1
```

**Environment (please complete the following information):**
 - Windows
 - Zephyr SDK and platformIO (zephyr v3.2.0)
 - git describe: zephyr-v3.2.0-3013-ge4fcb32451

**Additional context**

**Comment by henrikbrixandersen** [comment]
Can you reproduce this on current Zephyr `main` without using PlatformIO?

**Comment by s163555** [comment]
> Can you reproduce this on current Zephyr `main` without using PlatformIO?

I have included the zephyr sdk in the post, and it is the same behavior

I have both the NUCLEO-F411RE board and the XNUCLEO-F411RE board, and it seems that the regular NUCLEO-F411RE can flash, so it might be specific to the WaveShare board

**Comment by erwango** [comment]
> I have both the NUCLEO-F411RE board and the XNUCLEO-F411RE board, and it seems that the regular NUCLEO-F411RE can flash, so it might be specific to the WaveShare board

Ok, I just found out that XNUCLEO-F411RE is actually not the usual, ST, NUCLEO-F411RE.
I'm converting this to a discussion as this is not impacting an upstream board.

**Comment by erwango** [comment]
Regarding your issue. An important point is that your board is actually different from ST nucleo_f411re, so it's possible that openocd flashing scripts don't work.

Can you try to add the following to boards's openocd.cfg file:
```
reset_config connect_assert_srst
```

**Reply by s163555** [reply]
Which file is it specifically? I added the line to the file in zephyr/boards/arm/nucleo_f411re/support/openocd.cfg:
![image](https://user-images.githubusercontent.com/72868007/210480111-7982173a-39c6-4818-80b6-c7a1d3b9c679.png)
However, it still failed to reset

**Reply by erwango** [reply]
Yes, this is the file I was talking about.Thanks for having had a try.

If not working, I propose to use stm32cubeprogrammer as the runner for west flash.
In board's `board.cmake`, at the top of the file, add:
```
board_runner_args(stm32cubeprogrammer "--port=swd" "--reset=hw")
include(${ZEPHYR_BASE}/boards/common/stm32cubeprogrammer.board.cmake)
```

**Reply by s163555** [reply]
> Yes, this is the file I was talking about.Thanks for having had a try.
> 
> If not working, I propose to use stm32cubeprogrammer as the runner for west flash. In board's `board.cmake`, at the top of the file, add:
> 
> ```
> board_runner_args(stm32cubeprogrammer "--port=swd" "--reset=hw")
> include(${ZEPHYR_BASE}/boards/common/stm32cubeprogrammer.board.cmake)
> ```

It seems that the flashing is succesfully completed. 
![image](https://user-images.githubusercontent.com/72868007/210521683-c4cf3dbf-8873-46bb-bb2e-cf7c1d5f6a4c.png)
However, the same blinky/hello_world code that runs on nucleo_f411re does not work.
![image](https://user-images.githubusercontent.com/72868007/210522385-649e0c57-717e-412f-be7a-72c84f9dfd2e.png)

Regarding the blink, the LED on PA_5 is the same pin D13 as on nucleo_f411re?
![image](https://user-images.githubusercontent.com/72868007/210523324-9463551a-39bd-413a-a918-3a4d486eb6a9.png)

And regarding serial I can see the uart out is supposed to use uart2 (RX=PA_3, TX=PA_2) to send serial to the host machine through the on-board CP2102 IC. Does the standard nucleo_f411re use uart1?

Interestingly the board works interchangeably in mbed with blink and uart features with nucleo_f411re

**Reply by erwango** [reply]
You're using nucleo_f411re board definition with a board which is not an actual nucleo_f411re. You should expect some differences of behavior. Up to you to make the required changes that fit your board.

**Reply by s163555** [reply]
> You're using nucleo_f411re board definition with a board which is not an actual nucleo_f411re. You should expect some differences of behavior. Up to you to make the required changes that fit your board.

You are right. Sorry, I am just a total beginner so I am asking how to proceed. As far as I can see in the `zephyr/boards/arm/nucleo_f411re/nucleo_f411re.dts` file the pin definitions are correct and interchangeable between the two boards

**Comment by gabohc** [comment]
For me, it looks like I can't physically reset the board with NRESET.

Can you try this as well? Setting NRESET to ground to see if it works

**Reply by s163555** [reply]
> For me, it looks like I can't physically reset the board with NRESET.

This seems to be the case. For some reason `STM32CubeProgrammer` can reset the board but `west` cannot. 
> Can you try this as well? Setting `NRESET` to ground to see if it works

I tried adding a jumper between `GND` and `NRST` (PA13) and flashing again, but `west` still fails to flash the board with the same error.
```
Error: timed out while waiting for target halted
TARGET: stm32f3x.cpu - Not halted
```
Also, while PA13 is tied to `GND`, STM32CubeProgrammer gives the following error when connecting with the STLink v2 mini:
![image](https://user-images.githubusercontent.com/72868007/227867432-c2221491-6181-4085-9e37-d5bf967bf0c2.png)

While I can succesfully flash the board with the trick of hold and release of the reset button, I have not been able to get zephyr running. I have tried a blinky sample and an uart echo bot that both work on the NUCLEO-F411RE. I would be thrilled if this board could be supported by zephyr since it offers several improvements to the regular board.

**Reply by erwango** [reply]
> TARGET: stm32f3x.cpu - Not halted

There's an issue with the openocd configuration script you're using. Target should be 'stm32f4x.cpu'.
Also, to get Zephyr successfully running on the board, I suggest to look at the differences between the 2 boards (clocks, pins, ..) and update board .dts accordingly.

**Reply by s163555** [reply]
> > TARGET: stm32f3x.cpu - Not halted
> 
> There's an issue with the openocd configuration script you're using. Target should be 'stm32f4x.cpu'. Also, to get Zephyr successfully running on the board, I suggest to look at the differences between the 2 boards (clocks, pins, ..) and update board .dts accordingly.

Good spotted. My bad, my cmake cache was still building for a different board.
After reconfiguring to nucleo_f411re the corresponding script is used.
```
Info : [stm32f4x.cpu] Cortex-M4 r0p1 processor detected
Info : [stm32f4x.cpu] target has 6 breakpoints, 4 watchpoints
Info : starting gdb server for stm32f4x.cpu on 3333
Info : Listening on port 3333 for gdb connections
    TargetName         Type       Endian TapName            State
--  ------------------ ---------- ------ ------------------ ------------
 0* stm32f4x.cpu       hla_target little stm32f4x.cpu       running

Info : Unable to match requested speed 2000 kHz, using 1800 kHz
Info : Unable to match requested speed 2000 kHz, using 1800 kHz
Error: timed out while waiting for target halted
TARGET: stm32f4x.cpu - Not halted
```

**Reply by s163555** [reply]
When holding down the reset button and releasing it during the flashing process, it successfully flashes as such:
```
Info : The selected transport took over low-level target control. The results might differ compared to plain JTAG/SWD
Info : clock speed 2000 kHz
Info : STLINK V2J40S7 (API v2) VID:PID 0483:3748
Info : Target voltage: 3.293583
Info : [stm32f4x.cpu] Cortex-M4 r0p1 processor detected
Info : [stm32f4x.cpu] target has 6 breakpoints, 4 watchpoints
Info : starting gdb server for stm32f4x.cpu on 3333
Info : Listening on port 3333 for gdb connections
    TargetName         Type       Endian TapName            State
--  ------------------ ---------- ------ ------------------ ------------
 0* stm32f4x.cpu       hla_target little stm32f4x.cpu       reset

Info : Unable to match requested speed 2000 kHz, using 1800 kHz
Info : Unable to match requested speed 2000 kHz, using 1800 kHz
[stm32f4x.cpu] halted due to debug-request, current mode: Thread
xPSR: 0x01000000 pc: 0x080017c0 msp: 0x20001180
Info : Unable to match requested speed 8000 kHz, using 4000 kHz
Info : Unable to match requested speed 8000 kHz, using 4000 kHz
Info : device id = 0x10006431
Info : flash size = 512 KiB
auto erase enabled
wrote 16384 bytes from file C:/Users/Jeppe/zephyrproject/zephyr/samples/hello_world/build/zephyr/zephyr.hex in 0.618632s (25.864 KiB/s)

Info : Unable to match requested speed 2000 kHz, using 1800 kHz
Info : Unable to match requested speed 2000 kHz, using 1800 kHz
[stm32f4x.cpu] halted due to breakpoint, current mode: Thread
xPSR: 0x61000000 pc: 0x20000044 msp: 0x20001180
shutdown command invoked
```

**Reply by s163555** [reply]
I can also successfully verify the flashed device with the `zephyr.hex` file
![image](https://user-images.githubusercontent.com/72868007/227888522-f827d040-9cc4-4ec2-8115-4eb16f24e130.png)

**Comment by gabohc** [comment]
Hi @s163555 

Just a follow up question from someone experiencing similar issues.

Were you finally able to run any of the samples (Blinky) into the XNUCLEO F411RE? 

For me it doesn't with Zephyr. But if use the Blinky sample for the original NUCLEO F411RE from Mbed IOS, compiled with Keil Studio Cloud, and flashed with the STM32 ST-Link Utility, it does actually work

**Reply by s163555** [reply]
Hi gabohc,
I recognise your situation. The board was confirmed to work fine using Mbed and STM32 Cube toolchain. Unfortunately, I never managed to get zephyr to properly deploy on that board. It has several hardware improvements compared to the regular dev board, but it seems to be incompatible with the zephyr sdk.

**Reply by jurenat** [reply]
Hi, I have checked the [dts file](https://github.com/zephyrproject-rtos/zephyr/blob/f2e3d3f951112ffec7f3a1a40bcf0c187dcae40e/boards/st/nucleo_f411re/nucleo_f411re.dts#L53) and the clock tree is configured to use clock from ST link in bypass mode. STlink is not present on XNucleo (and MCO) so you should remove it from dts.

Also, NRST is not connected to the debug port. You can use an extra wire to connect it to the debug probe and it could start working.

**Reply by gabohc** [reply]
Hi darkmoon32

It worked! Removing the `hse-bypass;`  Great catch on that, how did you get that? 

Only referring to the compiled binary, I did not check on the flashing as for that I am using another method

Thank you very much! @s163555 Could you check if it works for you as well?

**Reply by s163555** [reply]
I will try later when I am off work.
@gabohc Just to clarify, does the change mean you can succesfully flash the board without manually resetting? Are you able to run a blinky example on the board? I was never able to do either.

**Reply by gabohc** [reply]
What it means is that I was able to run the blinky example. I compiled it with Zephyr and then loaded it with the ST Utility.  Haven't had the time yet to check further on the flashing

**Comment by ajarmouni-st** [comment]
> It worked! Removing the `hse-bypass;`

Hi @gabohc, great work! Would you be interested in submitting a PR that adds support for XNUCLEO F411RE? Here are some helpful resources in case you did decide to go for it https://docs.zephyrproject.org/latest/hardware/porting/board_porting.html https://docs.zephyrproject.org/latest/contribute/guidelines.html#contribute-guidelines

**Reply by gabohc** [reply]
I am interested! Will check on that
